tests/scripts/depends.py

```diff
@@ -17,9 +17,53 @@
 #
 # This file is part of Mbed TLS (https://tls.mbed.org)
 
-"""Test Mbed TLS with a subset of algorithms.
 """
+Test Mbed TLS with a subset of algorithms.
+
+This script can be divided into several steps:
+
+First, include/mbedtls/mbedtls_config.h or a different config file passed
+in the arguments is parsed to extract any configuration options (collect_config_symbols).
+
+Then, test domains (groups of jobs, tests) are built based on predefined data
+collected in the DomainData class. Here, each domain has five major traits:
+- domain name, can be used to run only specific tests via commandline;
+- configuration building method, described in detail below;
+- list of symbols passed to the configuration building method;
+- commands to be run on each job (only build, build and test, or any other custom);
+- optional list of symbols to be excluded from testing.
+
+The configuration building method can be one of the three following:
+
+- ComplementaryDomain - build a job for each passed symbol by disabling a single
```

Comment:
Design question: do we really need ComplementrayDomain / DualDomain at all?

We had a discussion about it for curves (which we moved form DualDomain to ExclusiveDomain), but I don't think any of the arguments were specific to curves, so perhaps everything should just use ExclusiveDomain? IIRC the main argument was that Complementary will only catch things that Exclusive missed in cases where something depends on at least 3 elements from the domain and at least 2 of them are declared (so Exclusive will always skip) but at least 1 isn't. That seemed like a rare enough occurrence that the additional testing time wasn't worth it.

Wdyt? cc @gilles-peskine-arm

Reply:
Agreed, I think we don't need ComplementaryDomain and DualDomain at all. They're there because when I started this script, we had realized ExclusiveDomain was catching actual problems that ComplementaryDomain didn't, but we hadn't done this reflection about ComplementaryDomain being largely redundant yet.

Reply:
Yes, fully agreed with Gilles, let's not delay this PR with improvements that can be done later. I didn't anticipate that removing ComplementaryDomain would create issue, but since it does, let's postpone it.

What really matters to unblock other tasks is the fixes for the issues found in cipher dependencies.